workflow/validate/validate.go

```diff
@@ -1290,21 +1334,51 @@ func validateDAGTargets(tmpl *wfv1.Template, nameToTask map[string]wfv1.DAGTask)
 	return nil
 }
 
-// verifyNoCycles verifies there are no cycles in the DAG graph
-func verifyNoCycles(tmpl *wfv1.Template, ctx *dagValidationContext) error {
+// verifyDAGNoCycles verifies there are no cycles in the DAG graph
+func verifyDAGNoCycles(tmpl *wfv1.Template, ctx *dagValidationContext) error {
+	depGraph := make(map[string][]string)
+	for _, task := range tmpl.DAG.Tasks {
+		depGraph[task.Name] = ctx.GetTaskDependencies(task.Name)
+	}
+	err := verifyNoCycles(depGraph)
+	if err != nil {
+		return errors.Errorf(errors.CodeBadRequest, "templates.%s.tasks %s", tmpl.Name, err.Error())
+	}
+	return nil
+}
+
+// verifyContainerSetNoCycles verifies there are no cycles in the ContainerSet dependency graph
+func verifyContainerSetNoCycles(tmpl *wfv1.Template) error {
+	depGraph := make(map[string][]string)
+	for _, ctr := range tmpl.ContainerSet.Containers {
+		depGraph[ctr.Name] = append(depGraph[ctr.Name], ctr.Dependencies...)
+	}
+	err := verifyNoCycles(depGraph)
+	if err != nil {
+		return errors.Errorf(errors.CodeBadRequest, "templates.%s.containerSet.containers %s", tmpl.Name, err.Error())
+	}
+	return nil
+}
+
+// verifyNoCycles verifies that a dependency graph has no cycles by Depth-First Search
+// depGraph is an adjacency list, where key is a node name and value is a list of its dependencies' names
```

Comment:
Had to refactor this func a little bit so that it can be shared between `DAG` and `ContainerSet`.

Reply:
Nice